master/static/srv/allocation_raw.sql

```diff
@@ -0,0 +1,79 @@
+WITH const AS (
+    SELECT
+        $1 :: timestamp AS day_start,
+        $2 :: timestamp AS day_end
+),
+-- Workloads that had any overlap with the target interval, with their start and end times bounded
+-- to the interval.
+workloads AS (
+    SELECT
+        all_workloads.trial_id,
+        all_workloads.kind,
+        all_workloads.start_time,
+        all_workloads.end_time,
+        extract(
+            epoch
+            FROM
+                CASE
+                    WHEN all_workloads.end_time IS NULL THEN const.day_end
+                    WHEN all_workloads.end_time > const.day_end THEN const.day_end
+                    ELSE all_workloads.end_time
+                END - CASE
+                    WHEN all_workloads.start_time < const.day_start THEN const.day_start
+                    ELSE all_workloads.start_time
+                END
+        ) AS seconds
+    FROM
+        (
+            -- Summarize the common relevant fields from all workload types. We might want this to
+            -- be a CTE, but I think that would cause PostgreSQL <12 to insert an optimization fence
+            -- and have to fully scan all three tables, which could be bad.
+            SELECT
+                'step' AS kind,
+                trial_id,
+                start_time,
+                end_time
+            FROM
+                steps
+            UNION ALL
+            SELECT
+                'validation' AS kind,
+                trial_id,
+                start_time,
+                end_time
+            FROM
+                validations
+            UNION ALL
+            SELECT
+                'checkpoint' AS kind,
+                trial_id,
+                start_time,
+                end_time
+            FROM
+                checkpoints
+        ) AS all_workloads,
+        const
+    WHERE
+        all_workloads.start_time <= const.day_end
```

Comment:
Yeah, I think we should do `all_workloads.end_time IS NOT NULL AND all_workloads.end_time > const.day_start AND all_workloads.end_time <= const.day_end` or else don't we lose accounting when workloads span days?

Reply:
I think the condition you're suggesting would exclude workloads that started during or before the period but ended after it, which we don't want.

Reply:
Yeah, it would but then if you query day by day you won't have any workloads not included in this aggregation (those you mention would get included in the next day). With it as is, if you run it daily, you'll miss information on workloads that don't fit into the interval [const.day_start, const.day_end) repeated over days (so workloads that start and end are in different days). If that is what this endpoint is for, this makes sense, but if someone is going to be calling if with [01-01-2020 00:00:00, 01-02-2020 00:00:00), [01-02-2020 00:00:00, 01-03-2020 00:00:00), etc.. this is going to miss stuff (e.g. a workload that spanned [01-01-2020 23:59:59, 01-02-2020 00:00:01] would never get included I think).

Reply:
No, that workload would be counted in _both_ days, contributing one second to each. The start time is before the ends of both periods and the end time is after the beginning of both periods.

Reply:
til i'm an idiot, missed the code where you only count the portion of the interval overlapping the queried.